### PR TITLE
Parallel installation of Xcode

### PR DESCRIPTION
**Pull request by dmitry-shibanov**
# Description
In scope of this pull request we add logic for parallel installation of Xcode's versions.

#### Related issue: https://github.com/actions/virtual-environments-internal/issues/1394

From schedule [build](https://github.visualstudio.com/virtual-environments/_build/results?buildId=95609&view=results):
- 03:17:34 | ./provision/core/xcode.ps1

With parallelisation [build](https://github.visualstudio.com/virtual-environments/_build/results?buildId=95350&view=results):
- 02:19:37 | ./provision/core/xcode.ps1

## Check list
- [x] Related issue / work item is attached
- [x] Tests are written (if applicable)
- [x] Documentation is updated (if applicable)
- [ ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/macos/helpers/Xcode.Installer.psm1 [hunk 1]
```diff
@@ -30,7 +27,9 @@ function Invoke-DownloadXcodeArchive {
 
     # TO-DO: Consider replacing of xcversion with own implementation
     Write-Host "Downloading Xcode $resolvedVersion"
-    Invoke-ValidateCommand "xcversion install '$resolvedVersion' --no-install"
+    Invoke-ValidateCommand "xcversion install '$resolvedVersion' --no-install" | out-null
```
Comment (maxim-lobanov) [inline images/macos/helpers/Xcode.Installer.psm1:30]: `Out-Host`

Comment (dmitry-shibanov) [reply images/macos/helpers/Xcode.Installer.psm1]: Done.

File: images/macos/helpers/Xcode.Installer.psm1 [hunk 2]
```diff
@@ -30,7 +31,13 @@ function Invoke-DownloadXcodeArchive {
 
     # TO-DO: Consider replacing of xcversion with own implementation
     Write-Host "Downloading Xcode $resolvedVersion"
-    Invoke-ValidateCommand "xcversion install '$resolvedVersion' --no-install"
+    Invoke-ValidateCommand "xcversion install '$resolvedVersion' --no-install" | out-host
```
Comment (maxim-lobanov) [inline images/macos/helpers/Xcode.Installer.psm1:34]: nitpick: `Out-Host` Please keep case

File: images/macos/helpers/Xcode.Installer.psm1 [hunk 3]
```diff
@@ -30,7 +31,13 @@ function Invoke-DownloadXcodeArchive {
...
+    $fixedResolvedVersion = "$resolvedVersion" -replace " ", "_"
```
Comment (maxim-lobanov) [inline images/macos/helpers/Xcode.Installer.psm1:35]: `$fixedResolvedVersion` is not clear variable name.
May be `$xcodeXipName`?

Comment (maxim-lobanov) [reply images/macos/helpers/Xcode.Installer.psm1:35]: `$xcodeXip` -> `$xcodeXipFile`

Comment (maxim-lobanov) [reply images/macos/helpers/Xcode.Installer.psm1:35]: `$tempXipDir` -> `$tempXipDirectory`

File: images/macos/helpers/Xcode.Installer.psm1 [hunk 4]
```diff
@@ -30,7 +31,13 @@ function Invoke-DownloadXcodeArchive {
...
+    $xcodeXip = Get-ChildItem -Path $DownloadDirectory -Filter "Xcode_$fixedResolvedVersion.xip" | Select-Object -First 1
+    $tempXipDir = New-Item -Path $DownloadDirectory -Name "Xcode$fixedResolvedVersion" -ItemType "directory"
```
Comment (maxim-lobanov) [inline images/macos/helpers/Xcode.Installer.psm1:37]: style nitpick: `directory` -> `Directory`

File: images/macos/helpers/Xcode.Installer.psm1 [hunk 5]
```diff
@@ -30,7 +31,13 @@ function Invoke-DownloadXcodeArchive {
...
+    Move-Item -Path "$xcodeXip" -Destination $tempXipDir
+
+    return "$tempXipDir"
```
Comment (maxim-lobanov) [inline images/macos/helpers/Xcode.Installer.psm1:40]: Why do we need quotes here and in 38 lines?

Comment (dmitry-shibanov) [reply images/macos/helpers/Xcode.Installer.psm1]: Done.

File: images/macos/helpers/Xcode.Installer.psm1 (hunk 4 above)
Comment (dmitry-shibanov) [reply images/macos/helpers/Xcode.Installer.psm1]: Done.

File: images/macos/helpers/Xcode.Installer.psm1 (hunk 3 above)
Comment (dmitry-shibanov) [reply images/macos/helpers/Xcode.Installer.psm1]: Done.

File: images/macos/helpers/Xcode.Installer.psm1 (hunk 2 above)
Comment (dmitry-shibanov) [reply images/macos/helpers/Xcode.Installer.psm1]: Done.

**Review by al-cheb — APPROVED** [review]

**Review by maxim-lobanov — APPROVED** [review]

File: images/macos/provision/core/xcode.ps1 [hunk 6]
```diff
@@ -14,20 +14,32 @@ $os = Get-OSVersion
...
 [Array]::Reverse($xcodeVersions)
+$threadCount = [Environment]::ProcessorCount
 
+Write-Host "ProcessorCount is $threadCount"
 Write-Host "Installing Xcode versions..."
-$xcodeVersions | ForEach-Object {
+$xcodeVersions | ForEach-Object -ThrottleLimit $threadCount -Parallel {
+    Import-Module "$env:HOME/image-generation/helpers/Common.Helpers.psm1"
+    Import-Module "$env:HOME/image-generation/helpers/Xcode.Installer.psm1"
+    $ErrorActionPreference = "SilentlyContinue"
```
Comment (maxim-lobanov) [inline images/macos/provision/core/xcode.ps1:24]: Why don't we use `$ErrorActionPreference = "Stop"` here?

File: images/macos/helpers/Xcode.Installer.psm1 [hunk 7]
```diff
@@ -30,7 +31,16 @@ function Invoke-DownloadXcodeArchive {
 
     # TO-DO: Consider replacing of xcversion with own implementation
     Write-Host "Downloading Xcode $resolvedVersion"
-    Invoke-ValidateCommand "xcversion install '$resolvedVersion' --no-install"
+    Invoke-ValidateCommand "xcversion install '$resolvedVersion' --no-install" -ErrorVariable xcversionError | Out-Host
+    if ($xcversionError){
+        Write-Host "Xcode $Version errors are $xcversionError"
```
Comment (maxim-lobanov) [inline images/macos/helpers/Xcode.Installer.psm1:36]: Change error message: `"xcversion install" command throws errors: $xcversionError`

**Review by maxim-lobanov — APPROVED** [review]

**Pull request merged**